mathesar/serializers.py

```diff
@@ -54,11 +54,54 @@ class SimpleColumnSerializer(serializers.Serializer):
 
 
 class ColumnSerializer(SimpleColumnSerializer):
-    index = serializers.IntegerField(source='column_index', read_only=True)
+    name = serializers.CharField(required=False)
+
+    # From scratch fields
+    type = serializers.CharField(source='plain_type', required=False)
     nullable = serializers.BooleanField(default=True)
     primary_key = serializers.BooleanField(default=False)
+
+    # From duplication fields
+    source_column = serializers.IntegerField(required=False, write_only=True)
+    copy_source_data = serializers.BooleanField(default=True, write_only=True)
+    copy_source_constraints = serializers.BooleanField(default=True, write_only=True)
+
+    # Read only fields
+    index = serializers.IntegerField(source='column_index', read_only=True)
     valid_target_types = serializers.ListField(read_only=True)
 
+    def validate(self, data):
+        if not self.partial:
+            from_scratch_required_fields = ['name', 'type']
+            from_scratch_specific_fields = ['type', 'nullable', 'primary_key']
+            from_dupe_required_fields = ['source_column']
+            from_dupe_specific_fields = ['source_column', 'copy_source_data',
+                                         'copy_source_constraints']
+
+            # Note that we run validation on self.initial_data, as `data` has defaults
+            # filled in for fields that weren't specified by the request
+            from_scratch_required_all = all([
+                f in self.initial_data for f in from_scratch_required_fields
+            ])
+            from_scratch_specific_in = [
+                f for f in from_scratch_specific_fields if f in self.initial_data
+            ]
+            from_dupe_required_all = all([
+                f in self.initial_data for f in from_dupe_required_fields
+            ])
+            from_dupe_specific_in = [
+                f for f in from_dupe_specific_fields if f in self.initial_data
+            ]
+
+            if len(from_dupe_specific_in) and len(from_scratch_specific_in):
+                raise ValidationError(
+                    f'{from_scratch_specific_in} cannot be passed in if '
+                    f'{from_dupe_specific_in} has also been passed in.'
+                )
+            elif not from_dupe_required_all and not from_scratch_required_all:
+                raise ValidationError('Required fields not found.')
```

Comment:
Please ensure this error matches the default DRF error for required fields not being passed in. I believe default DRF errors would return the field as the key and the error message as the value. See: https://www.django-rest-framework.org/api-guide/exceptions/

Reply:
Given we have two distinct sets of required fields, what would this look like? Doing something like
```python
{
  'type': ['This field is required.'],
  'name': ['This field is required.'],
  'source_column': ['This field is required.'],
}
```
seems like it would be confusing (especially since name is optional for duplication). Would it make sense to have column from scratch as the "default" path, and return just
```python
{
  'type': ['This field is required.'],
  'name': ['This field is required.'],
}
```

Reply:
You'd want to return an appropriate error based on what's passed in. e.g. if `copy_source_data` was passed in, you'd want to only return `'source_column': ['This field is required.'],`. if `type` was passed in, you'd want to return `'name': ['This field is required.'],`. You should be able to figure this out based on the required/specific fields you've already calculated.

Separately, we shouldn't require name for column creation, we could auto-generate one like we do for tables. If you don't want to handle that in this PR, please make a separate issue for it.

Reply:
Updated the error, with it defaulting to from scratch if no parameters were passed. Also opened #555.